### ISSUE TITLE
[tt_metal/hw][SFPU] 3 numerical-correctness defects: isclose bit-pattern compare (WH+BH), Quasar exp missing range guards

### ISSUE THREAD

**Issue by amahmudTT** [tenstorrent/tt-metal#51220]
These are statically audited bugs. The issues have enough detail that you can point claude and ask it to fix them. There is a small chance of false positives.

---

Three high-severity numerical defects in the SFPU kernels under `tt_metal/hw/ckernels/*/metal/llk_api/llk_sfpu/`. Each returns a silently wrong value on a supported input range — no assert, no NaN, just the wrong number.

**Line references are against `02246470f99`, and every file below was re-checked as byte-identical on `main` @ `007e45f8bfe`, so the line numbers are valid as of filing. If lines have since moved, the quoted code in each entry is the reliable anchor.**

Each item is self-contained — location, triggering configuration, evidence, and a suggested fix — and is independently fixable. Please tick items off as they land rather than closing the issue on the first fix.

CODEOWNERS: @rtawfik01 @rdjogoTT @nvelickovicTT

**Items 1 and 2 are the same defect in two architecture copies of the same file**, at the same line, so fix them together. Item 3 is a Quasar port that dropped guards its Blackhole source has.

Note on verification: all three are numerically checkable, but the behaviour is architecture-specific — item 3 needs Quasar, items 1-2 need WH and BH respectively.

| # | Sev | Location | Defect |
|---|---|---|---|
| 1 | H | [`tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h:81`](https://github.com/tenstorrent/tt-metal/blob/007e45f8bfe/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h#L81) | isclose's Inf/NaN fix-up compares the FLOAT magnitudes a_abs/b_abs against the integer bit-pattern constant inf_bits (0x7F800000), so the comparison silently beco... |
| 2 | H | [`tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h:81`](https://github.com/tenstorrent/tt-metal/blob/007e45f8bfe/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h#L81) | The Inf/NaN fix-up in isclose compares the *float magnitude* of the operands against the integer bit pattern 0x7F800000, so the special-case branch fires for ever... |
| 3 | H | [`tt_metal/hw/ckernels/quasar/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h:91`](https://github.com/tenstorrent/tt-metal/blob/007e45f8bfe/tt_metal/hw/ckernels/quasar/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h#L91) | Quasar's fp32 accurate exp reconstructs 2^i with a raw `(i + 127) << 23` and drops the overflow/underflow guards the Blackhole source it was ported from has, so a... |

---

### 1. `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h:81` — wrong-comparison

**isclose's Inf/NaN fix-up compares the FLOAT magnitudes a_abs/b_abs against the integer bit-pattern constant inf_bits (0x7F800000), so the comparison silently becomes a float compare against 2139095040.0f instead of the intended bit compare; matching infinities return 0 and any finite |x| >= 2.139e9 is misclassified as a special value and forced to 0.**

**How it fails.** ttnn.isclose(a, b) with float32 inputs. (1) a = b = +inf: the fix-up branch is entered (inf >= 2139095040.0f is true), result is reset to 0, and the recovery test `a_abs == inf_bits` evaluates inf == 2139095040.0f => false, so the kernel writes 0 while torch.isclose(inf, inf) is True. (2) a = b = 3.0e9 (ordinary finite float32): `a_abs >= inf_bits` is 3.0e9 >= 2139095040.0f => true, so the lane is treated as special, result is zeroed, `a_abs == inf_bits` is false (3.0e9 != 2.139e9), and isclose returns 0 for two bit-identical equal values.

**Evidence.** Line 37 declares the value as an integer bit pattern: `constexpr int32_t inf_bits = 0x7F800000;` with the comment 'IEEE-754 abs(+inf). abs_bits == inf_bits -> +-Inf; abs_bits > inf_bits -> NaN'. The guard comment at line 73 also says 'Detected by `abs_bits >= inf_bits` which holds iff exp == 0xFF'. But the code compares floats, not bits:
```
sfpi::vFloat a_abs = sfpi::abs(a);          // line 53, vFloat (sfpi_lib.h: abs(vFloat) -> vFloat)
sfpi::vFloat b_abs = sfpi::abs(b);
...
v_if(a_abs >= inf_bits || b_abs >= inf_bits) {      // line 81
    result = 0.0f;
    v_if(a_abs == inf_bits && a_bits == b_bits) { result = 1.0f; }   // line 83
```
Overload resolution: sfpi has `operator>=(vFloat, float)` and `operator>=(vInt, int32_t)`; vInt has no converting constructor from vFloat (sfpi.h:235-251), so the only viable candidate is (vFloat, float) with a standard int32_t->float conversion. 0x7F800000 therefore becomes the finite value 2139095040.0f, not the Inf bit pattern. The intent to use bit views is visible in the code itself: `a_bits`/`b_bits` are computed at lines 51-52 and the header comment says 'Cache integer views of a, b and their abs bits up front ... These feed ... the merged Inf/NaN fix-up branch below', yet no abs-bits vInt is ever formed and `b_bits` is used nowhere except the `a_bits == b_bits` test. Existing tests cannot catch this: tests/ttnn/unit_tests/operations/eltwise/test_relational.py::test_isclose_inf_divergence only checks MISMATCHED infinities (expected False, which the broken code happens to produce), and every other isclose test uses ranges within +-150.

**Suggested fix.** Form the abs bit patterns as vInt and compare those, e.g. `sfpi::vInt a_abs_bits = a_bits & 0x7FFFFFFF; sfpi::vInt b_abs_bits = b_bits & 0x7FFFFFFF;` then `v_if(a_abs_bits >= inf_bits || b_abs_bits >= inf_bits)`, `v_if(a_abs_bits == inf_bits && a_bits == b_bits)`, and `v_if(a_abs_bits > inf_bits && b_abs_bits > inf_bits)`. Apply the identical change to the byte-identical Blackhole copy at tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h, and add regression cases for isclose(+inf, +inf) / isclose(-inf, -inf) and for equal finite values above 2.14e9.

<sub>CODEOWNERS: @rtawfik01 @rdjogoTT @nvelickovicTT</sub>

- [ ] fixed

---

### 2. `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_isclose.h:81` — wrong-cast

**The Inf/NaN fix-up in isclose compares the *float magnitude* of the operands against the integer bit pattern 0x7F800000, so the special-case branch fires for every finite input whose magnitude is >= 2139095040.0 and never fires for exact +-Inf equality; isclose(inf, inf) returns 0 and isclose(x, x) returns 0 for |x| >= 2.139e9.**

**How it fails.** a = b = +inf (fp32 or bf16 tensor): abs_diff = inf-inf = NaN so the tolerance test gives 0; the fix-up branch is entered (inf >= 2139095040.0f) but `a_abs == inf_bits` is `inf == 2139095040.0f` = false, so the result stays 0. torch.isclose(inf, inf) is True -> wrong output. Likewise a = b = 3e9 (finite): both operands enter the fix-up branch (3e9 >= 2139095040.0f), result is reset to 0 and neither rebuild condition matches, so isclose(3e9, 3e9) returns 0 instead of 1. Existing tests miss it: tests/ttnn/unit_tests/operations/eltwise/test_relational.py::test_isclose_inf_divergence only checks inf vs -inf (expected False), and all value-range tests stay well under 2e9.

**Evidence.** `constexpr int32_t inf_bits = 0x7F800000;` (line 37) is documented as a *bit pattern* ("abs_bits == inf_bits -> +-Inf; abs_bits > inf_bits -> NaN ... Detected by `abs_bits >= inf_bits` which holds iff exp == 0xFF"), but the operands it is compared against are floats, not bit views: `sfpi::vFloat a_abs = sfpi::abs(a);` (line 53; sfpi::abs(vFloat) returns vFloat per runtime/sfpi/include/sfpi_lib.h:280). sfpi declares only `operator>=(vFloat, vFloat)` and `operator>=(vFloat, float)` (sfpi.h:490/498); vFloat has no implicit conversion to vInt (class vInt has no vFloat ctor, and impl_::vVal exposes get() only, no conversion operator), so `a_abs >= inf_bits` / `a_abs == inf_bits` / `a_abs > inf_bits` (lines 81, 83, 86) bind to the float overload with the integer converted to 2139095040.0f -- a plain magnitude comparison, not an exponent test. The correct idiom is used by the sibling kernel in the same directory: ckernel_sfpu_log1p.h:123 `v_if(sfpi::as<sfpi::vInt>(u) >= sfpi::as<sfpi::vInt>(infinity))`.

**Suggested fix.** Compare integer views: keep the already-computed `a_bits`/`b_bits`, derive `a_abs_bits = a_bits & 0x7FFFFFFF` (same for b), and use those in the three special-case predicates; keep the float `b_abs` only for the tolerance term. Add regression cases for isclose(inf, inf) and for |x| ~ 1e10.

<sub>CODEOWNERS: @rtawfik01 @rdjogoTT @nvelickovicTT</sub>

- [ ] fixed

---

### 3. `tt_metal/hw/ckernels/quasar/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h:91` — correctness

**Quasar's fp32 accurate exp reconstructs 2^i with a raw `(i + 127) << 23` and drops the overflow/underflow guards the Blackhole source it was ported from has, so any input outside roughly (-88.03, +88.72) yields a wrong-signed, wildly wrong result instead of 0 / +inf.**

**How it fails.** exp_tile<approx=false>() on Quasar with a 32-bit Dest (DST_ACCUM_MODE=true) routes every lane to _sfpu_exp_fp32_accurate_ (there is no input-magnitude routing in calculate_exponential, only the EN_32BIT_DEST/APPROXIMATION_MODE switch). For a = -100.0f: j = round(1.442695*a) = -144, so i = -144, i+127 = -17. (-17) two's-complement is 0xFFFFFFEF; `<< 23` keeps only bits 0..8 (0x1EF) giving 0xF7800000, i.e. -2^112. With r = exp(f) ~= 0.8296 the function returns about -4.3e33 instead of exp(-100) ~= 3.7e-44 (~0). Symmetrically a = +100.0f gives i+127 = 271 = 0x10F, `<< 23` -> 0x87800000 = -2^-112, i.e. about -1.6e-34 instead of +inf. This is the softmax path: masked attention logits (x - max = -1e4 or -1e9) go through exp and come back as huge NEGATIVE numbers, poisoning the sum and every probability in the row. For a magnitude beyond 2^22 the helper's own documented domain is also violated, so the reduced argument is garbage too.

**Evidence.** Quasar (this file, lines 81-92): the entire reconstruction is
    sfpi::vFloat two_i = sfpi::as<sfpi::vFloat>((i + 127) << 23);
    return r * two_i;
with the comment 'Correct across fp32 exp's representable domain (|a| < ~88); larger-magnitude inputs are the LUT/approx path's concern.' That justification is contradicted by the dispatcher 20 lines below (lines 118-122):
    if constexpr (!EN_32BIT_DEST || APPROXIMATION_MODE) { result = sfpi::approx_exp(val); } else { result = _sfpu_exp_fp32_accurate_(val); }
-- the selection is purely compile-time on EN_32BIT_DEST/APPROXIMATION_MODE; nothing ever routes a large-magnitude *value* to the LUT path.
The Blackhole source it was ported from (tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h, _sfpu_exp_fp32_accurate_<unsafe=false>, lines ~375-395) explicitly handles both ends:
    y *= std::numeric_limits<float>::infinity();            // overflow -> inf/NaN
    e = sfpi::exexp(r, sfpi::ExponentMode::Biased) + i;
    v_block { vInt e_lt_255 = __builtin_rvtt_sfpiadd_i(e.get(), -255, SFPIADD_MOD1_CC_LT0);
              y = sfpi::setexp(r, e);
              v_if (e_lt_255 < -254) { y = 0.0f; } v_endif; } v_endblock;
and the unguarded variant is opt-in only via the `unsafe` template parameter (_sfpu_exp_fp32_accurate_unsafe_). Blackhole's bf16 path likewise clamps (`xlog2 = sfpi::clamp(xlog2, 0.0f, 255.0f)`), so on BH both paths are range-safe.
The Quasar entry point additionally advertises the clamp it never performs: calculate_exponential takes CLAMP_NEGATIVE and does `static_assert(CLAMP_NEGATIVE == true, ...)` (line 109) while marking it [[maybe_unused]], and the public API (tt_metal/hw/inc/api/compute/eltwise_unary/exp.h) documents InputClamping::ClampToNegative as 'Inputs below ~-88.5 are clamped to -88.5' and passes it straight through.

**Suggested fix.** Port the Blackhole guards: build 2^i with exexp/setexp and add the two range checks (result 0.0f when the biased exponent underflows, +inf when it overflows), or -- cheaper and consistent with the CLAMP_NEGATIVE contract already static_asserted true -- clamp the input to [-88.5f, 88.5f] (e.g. sfpi::clamp) before the range reduction in calculate_exponential. Either way remove the 'larger-magnitude inputs are the LUT/approx path's concern' comment, since no such routing exists.

<sub>CODEOWNERS: @rtawfik01 @rdjogoTT @nvelickovicTT</sub>

- [ ] fixed

---
